### PR TITLE
MXBackgroundSyncService: Keep its stores in sync with MXSession store

### PR DESCRIPTION
**Pull request by manuroe**
In real life, the same `MXBackgroundSyncService` instance runs for hours.  `MXSession` can run in parallel with it.

Here, we make sure that `MXBackgroundSyncService` uses the latest data stored by `MXSession`, instead of using obsolete data we have in memory.
Else `MXSession` will consider `MXBackgroundSyncService` data obsolete on its next startup or resume, making `MXBackgroundSyncService` previous work useless.

MXBackgroundStore.event(withEventId:) was not working. It should work now.

### CODE REVIEW

File: MatrixSDK/Background/MXBackgroundSyncService.swift [hunk 1]
```diff
@@ -144,6 +143,11 @@ public enum MXBackgroundSyncServiceError: Error {
                         completion: @escaping (MXResponse<MXEvent>) -> Void) {
         NSLog("[MXBackgroundSyncService] fetchEvent: \(eventId). allowSync: \(allowSync)")
         
+        // Check local stores on every request so that we use up-to-data data from the MXSession store
+        if allowSync {
+            updateBackgroundServiceStoresIfNeeded()
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundSyncService.swift:148]: It may be a good idea to move this in `launchBackgroundSync`, as there are other cases triggering background sync.

File: MatrixSDK/Background/MXBackgroundSyncService.swift [hunk 2]
```diff
@@ -522,4 +527,24 @@ public enum MXBackgroundSyncServiceError: Error {
...
+        // by checking that the event stream token we have in memory is the same that in the last version of MXSession store
+        let eventStreamToken = store.eventStreamToken
+ 
+        let upToDateStore = MXBackgroundStore(withCredentials: credentials)
+        let upToDateEventStreamToken = upToDateStore.eventStreamToken
+
+        if (eventStreamToken != upToDateEventStreamToken) {
+            // MXSession continued to work in parallel with the background sync service
+            // MXSession has updated its stream token. We need to use t
+            NSLog("[MXBackgroundSyncService] resetBackgroundServiceStoresIfNeeded: Update MXBackgroundStore. Its event stream token (\(String(describing: eventStreamToken))) does not match current MXStore.eventStreamToken (\(String(describing: upToDateEventStreamToken)))")
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundSyncService.swift:541]:
```suggestion
            NSLog("[MXBackgroundSyncService] updateBackgroundServiceStoresIfNeeded: Update MXBackgroundStore. Its event stream token (\(String(describing: eventStreamToken))) does not match current MXStore.eventStreamToken (\(String(describing: upToDateEventStreamToken)))")
```

File: MatrixSDK/Background/MXBackgroundSyncService.swift [hunk 3]
```diff
@@ -522,4 +527,24 @@ public enum MXBackgroundSyncServiceError: Error {
...
+            store = upToDateStore
+            
+            // syncResponseStore has obsolete data. Reset it
+            NSLog("[MXBackgroundSyncService] resetBackgroundServiceStoresIfNeeded: Reset MXSyncResponseStore. Its prevBatch was token \(String(describing: syncResponseStore.prevBatch))")
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundSyncService.swift:545]:
```suggestion
            NSLog("[MXBackgroundSyncService] updateBackgroundServiceStoresIfNeeded: Reset MXSyncResponseStore. Its prevBatch was token \(String(describing: syncResponseStore.prevBatch))")
```

File: MatrixSDK/Background/MXBackgroundStore.swift [hunk 4]
```diff
@@ -83,45 +58,177 @@ class MXBackgroundStore: MXMemoryStore {
     }
     
     //  This store should act like as a permanent one
-    override var isPermanent: Bool {
+    var isPermanent: Bool {
         return true
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundStore.swift:62]:
```suggestion
        return false
```

I think we may stop acting like a permanent store, as we don't use a session anymore.

File: MatrixSDK/Background/MXBackgroundStore.swift [hunk 5]
```diff
@@ -83,45 +58,177 @@ class MXBackgroundStore: MXMemoryStore {
     }
     
     //  This store should act like as a permanent one
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundStore.swift:60]:
```suggestion
```

File: MatrixSDK/Background/MXBackgroundStore.swift [hunk 6]
```diff
@@ -83,45 +58,177 @@ class MXBackgroundStore: MXMemoryStore {
...
+    }
+    
+    func store(_ user: MXUser) {
+    }
+    
+    func users() -> [MXUser]? {
+        return nil
+    }
+    
+    func user(withUserId userId: String) -> MXUser? {
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundStore.swift:188]: We use the result of this function at `MXBackgroundSyncService.swift:132`. Why did you remove fetching our user from the store?

File: MatrixSDK/Background/MXBackgroundStore.swift [hunk 7]
```diff
@@ -24,57 +24,32 @@ enum MXBackgroundStoreErrorCode: Int {
...
-                                 userInfo: nil))
-                return
-            }
-            //  load session user before calling onComplete
-            self.fileStore.asyncUsers(withUserIds: [userId], success: { (users) in
-                if let user = users.first {
-                    self.myUser = user
-                }
-                onComplete?()
-            }, failure: failure)
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundStore.swift:58]: We need this, commented about at `user(withUserId userId: String) -> MXUser?` method.

File: MatrixSDK/Background/MXBackgroundStore.swift [hunk 8]
```diff
@@ -83,45 +58,177 @@ class MXBackgroundStore: MXMemoryStore {
...
+    }
+    
+    func storeEvent(forRoom roomId: String, event: MXEvent, direction: __MXTimelineDirection) {
+    }
+    
+    func replace(_ event: MXEvent, inRoom roomId: String) {
+    }
+    
+    func eventExists(withEventId eventId: String, inRoom roomId: String) -> Bool {
+        return false
```
Comment (ismailgulek) [inline MatrixSDK/Background/MXBackgroundStore.swift:136]: It'd be perfect to implement this in an efficient way in order to avoid reading whole events for only one. Not mandatory for now but it can be a good improvement point in future.

File: MatrixSDKTests/MXBackgroundSyncServiceTests.swift [hunk 9]
```diff
@@ -584,6 +584,163 @@ class MXBackgroundSyncServiceTests: XCTestCase {
...
+                        self.bgSyncService = MXBackgroundSyncService(withCredentials: bobCredentials)
+                        self.bgSyncService?.event(withEventId: eventId, inRoom: roomId) { _ in
+                            
+                            let syncResponseStore = MXSyncResponseFileStore()
+                            syncResponseStore.open(withCredentials: bobCredentials)
+                            let syncResponseStorePrevBatch = syncResponseStore.prevBatch
+                            
+                            // Test check
+                            // MXSyncResponseFileStore should point to the same sync token as MXSession store
+                            XCTAssertNotEqual(bobSessionStartEventStreamToken1, syncResponseStorePrevBatch)
```
Comment (ismailgulek) [inline MatrixSDKTests/MXBackgroundSyncServiceTests.swift:639]: This should be `XCTAssertEqual` in theory, but this is worthless. Because Bob's session is not stopped in this test, we cannot guarantee that his store's eventStreamToken will be same as sync response store's prevBatch.

**Review by ismailgulek — CHANGES_REQUESTED** [review]

File: MatrixSDK/Background/MXBackgroundStore.swift (hunk 7 above)
Comment (manuroe) [reply MatrixSDK/Background/MXBackgroundStore.swift:58]: I will check again why we need `user(withUserId userId: String) -> MXUser?`
My issue was this is an async call within a synchronous method. I saw cases where the code was doing things whereas the async call was not complete.

File: MatrixSDK/Background/MXBackgroundStore.swift (hunk 8 above)
Comment (manuroe) [reply MatrixSDK/Background/MXBackgroundStore.swift:135]: I do not think we use this method in the bg service. If I am wrong, we need to implement it.

File: MatrixSDK/Background/MXBackgroundStore.swift (hunk 6 above)
Comment (manuroe) [reply MatrixSDK/Background/MXBackgroundStore.swift:187]: Thanks for the catch. I missed it. I implemented `MXBackgroundSyncService.swift:132` in another way.

Comment (ismailgulek) [reply MatrixSDK/Background/MXBackgroundStore.swift:187]: Cool, thanks

File: MatrixSDK/Background/MXBackgroundStore.swift (hunk 8 above)
Comment (ismailgulek) [reply MatrixSDK/Background/MXBackgroundStore.swift:135]: No, you right, as we're not using it right now. Mine was just a wish.

File: MatrixSDKTests/MXBackgroundSyncServiceTests.swift [hunk 10]
```diff
@@ -584,6 +584,157 @@ class MXBackgroundSyncServiceTests: XCTestCase {
...
+                                    
+                                    // -> MXBackgroundSyncService should have detected that the MXSession ran in parallel.
+                                    //    It must have reset its cache. syncResponseStore.prevBatch must not be the same
+                                    let syncResponseStore = MXSyncResponseFileStore()
+                                    syncResponseStore.open(withCredentials: bobCredentials)
+                                    
+                                    XCTAssertNotEqual(syncResponseStorePrevBatch, syncResponseStore.prevBatch)
+                                    
+                                    // -> MXSyncResponseFileStore must be in sync with MXSession store
+                                    XCTAssertNotEqual(bobSessionStartEventStreamToken2, syncResponseStore.prevBatch)
```
Comment (ismailgulek) [inline MatrixSDKTests/MXBackgroundSyncServiceTests.swift:654]: Seems same with https://github.com/matrix-org/matrix-ios-sdk/pull/968#discussion_r538401678

File: MatrixSDK/Background/MXBackgroundSyncService.swift (hunk 1 above)
Comment (ismailgulek) [reply MatrixSDK/Background/MXBackgroundSyncService.swift:147]: > Do you think we need to call this method for every call of `launchBackgroundSync`?

Yes, just in case for parallel session syncs.

**Review by ismailgulek — APPROVED** [review]
